manticore/native/cpu/arm.py

```diff
@@ -20,35 +23,108 @@ def HighBit(n):
+    def decorator(body):
+        if can_take_denormalized_mod_imm:
+            # Need to possibly normalize a modified immediate argument that's
+            # been explicitly split by Capstone into (number, rotation)
+            # components.
+
+            body_sig = inspect_signature(body)
+            num_body_params = len(body_sig.parameters) - 1
+            assert num_body_params > 0
+            address_width = ARMV7_CPU_ADDRESS_BIT_SIZE
+
+            def normalize_mod_imm_arg(args):
+                if len(args) == num_body_params + 1:
+                    args = list(args)
+                    rot = args.pop()
+                    assert rot.type == "immediate"
+                    num = args.pop()
+                    assert num.type == "immediate"
+                    imm = ROR(num.imm, rot.imm, address_width)
+                    args.append(_ImmediatePseudoOperand(imm))
+                return args
 
-        if cpu._at_symbolic_conditional == cpu.instruction.address:
-            cpu._at_symbolic_conditional = None
-            should_execute = True
         else:
-            if issymbolic(should_execute):
-                # Let's remember next time we get here we should not do this again
-                cpu._at_symbolic_conditional = cpu.instruction.address
-                i_size = cpu.instruction.size
-                cpu.PC = Operators.ITEBV(
-                    cpu.address_bit_size, should_execute, cpu.PC - i_size, cpu.PC
-                )
-                return
+            # No normalization of a modified immediate required; return what's given.
+            def normalize_mod_imm_arg(args):
+                return args
 
-        if should_execute:
-            ret = body(cpu, *args, **kwargs)
+        @wraps(body)
+        def instruction_implementation(cpu, *args, **kwargs):
+            should_execute = cpu.should_execute_conditional()
 
-        if cpu.should_commit_flags():
-            cpu.commit_flags()
+            if cpu._at_symbolic_conditional == cpu.instruction.address:
+                cpu._at_symbolic_conditional = None
+                should_execute = True
+            else:
+                if issymbolic(should_execute):
+                    # Let's remember next time we get here we should not do this again
+                    cpu._at_symbolic_conditional = cpu.instruction.address
+                    i_size = cpu.instruction.size
+                    cpu.PC = Operators.ITEBV(
+                        cpu.address_bit_size, should_execute, cpu.PC - i_size, cpu.PC
+                    )
+                    return
+
+            if should_execute:
+                ret = body(cpu, *normalize_mod_imm_arg(args), **kwargs)
+            else:
+                ret = None
 
-        return ret
+            if cpu.should_commit_flags():
+                cpu.commit_flags()
+
+            return ret
+
+        return abstract_instruction(instruction_implementation)
+
+    if body is not None:
```

Comment:
`body` shouldn't ever be None, right? Isn't the first argument to a decorator always the wrapped function?

Reply:
That conditional at the end is to support using the decorator with 2 forms:
```
@instruction
def foo(...)
```
and this:
```
@instruction(can_take_denormalized_mod_imm=True)
def foo(...)
```
That is, allow using the decorator both with and without arguments.

In the first case, `body` will be bound to the function being decorated.  In the second case, however, `body` will be bound to the default value `None`.  If you were to write out the return type of `instruction`, it would be complicated, and would depend on exactly how it is called.

See https://stackoverflow.com/questions/3888158/making-decorators-with-optional-arguments for some more explanation.

I've expanded commentary on this decorator.

Reply:
Okay, so we've basically hacked support for two different function signatures in there. I'm glad mypy doesn't seem to mind.